### ISSUE TITLE
[DISCUSSION/FEATURE/REFACTOR] Decentralize IPC modules

### ISSUE THREAD

**Issue by nathanlesage** [Zettlr/Zettlr#859]
As I mentioned in the past, after the FSAL rewrite has been done, it's time to prepare the next big refactoring process. I personally think that after that one, the app is decentralized and modularized to a degree that scalability is once again reached sufficiently. (If I'm wrong and there are other modules that are difficult to scale, please feel free to have a look!)

## What's this about?

Basically, it concerns a full rewrite of the inter process communication logic, which is currently heavily centralized in two main classes: `zettlr-ipc.js` and `zettlr-rendereripc.js`. There are multiple problems with this.

One: When I began writing the app, I did not anticipate it would at some point include multiple windows, so the idea was to rout all communication to interface-like class objects that could do the heavy lifting so that from throughout both renderer and main process one could simply send messages and be sure they'd land on the other end. However, there are two problems:

1. Every message that arrives FROM the counterpart needs to be fed through the main object in each process, so even if a message is needed in the editor alone, it'll need to pass through the renderer object. The same holds true for the main process: Even if an object solely concerns the citations, it'll need to pass through the main process handler.
2. Additionally, this dependence on one specific class has led to a curious behaviour I recently noticed when I switched the behaviour of the `Close Window` menu item to `Close Tab, and close window if there are no more tabs`, as — no matter which window was actually focused — the message would always be deferred to the main window, where it would close a tab — even if your intention was to close the log window.

## How should we do it?

So here's my idea of what we might do, where your input is warmly welcome:

1. Create two providers (i.e. still have one object, but that object shall be completely agnostic to what is being passed to it), which'll mostly work as the current ones — with two specific changes: First, it will only take `on`-listeners, and diseminate the events to these without interfering, instead of trying to handle them itself, and second, it will allow either to target "the current focused" window or a specific one (at least in the main process, the renderer IPC calls will always be pointing towards the one single main process)
2. Create a sensible model of what kinds of channels we need to patch data through
3. During implementation, get rid of all of these multiple-messages being sent

### Communication Channels

So here's what I think we need concerning channels (if you are new to IPC, channels equal the different events an EventEmitter can emit, so we can think of any string we like)

* `command` (main only) --> Will reach the main application which will try to find commands that respond to the given command
* `citeproc` --> Will only facilitate communication between the renderer's editor and the sidebar, e.g. something encapsulated in the `citeproc-provider.js`
* `fsal` (renderer only) --> Anything file-related; this'll basically map the events from the FSAL to IPC calls send towards the renderer
* `tags` --> Everything tag-related
* `config` --> Simply used to transmit config values back and forth
* `dictionary` (better name?) --> For spell-checking purposes
* Which channels would be necessary as well, which I forgot here? Search, maybe …?

### Interface

Here's the first, raw idea of implementing the interface:

```javascript
// main process

// Sends command by default to the currently focused window, or a specific one
global.ipc.send(channel, command, payload, window = focusedWindow)

// Listens to a channel.
global.ipc.on(channel, callback(command, playload, originWindow) {
  /* originWindow can be used to immediately bounce back
     some result to the sender --> global.ipc.get, see below.
     What this means is that you can use originWindow.webContents.send
     to immediately return an (asynchronous) reply */
})

// Broadcasts something that should not be sent to one, but all windows
// at the same time, such as config changes. Think of this as EventEmitter.emit,
// only for BrowserWindow instances.
global.ipc.broadcast(channel, command, payload)

// renderer process

// "Fire and forget" a command w/ a payload to the main process
global.ipc.send(channel, command, payload)

// Returns a promise that will resolve with the answer from main, TODO: Naming?
global.ipc.get(channel, command, payload)

// Listens for messages (e.g. broadcasts)
global.ipc.on(channel, callback(command, payload) { /**/ })
```

As you can see the renderer process differs slightly: Whereas the main process allows you to select windows which shall receive your message, the renderer process may also use `get` as some form of "acknowledgement" callback. This will enable messages being sent, immediately evaluated and sent back. Basically, this corresponds to "synchronous" IPC calls (which we should not use, which is why I would like to port the current asynchronous callback algorithm which works pretty good).

Why doesn't the main process have the ability to acknowledge or demand immediate responses to its messages? Well, this has to do with where I want the source of truth to be. What matters is always the main state, so the renderer process needs to make sure to re-fetch anything from the "correct" state in the main process if something goes wrong, but the main process simply emanated changes and doesn't care whether or not renderer processes do something with this. This furthermore keeps functionality simple. This "acknowledgement" functionality is already being used by the citeproc handler (I think in the editor class somewhere) as well as the dictionary provider. We could extend this to include also other places where it would make sense to grab something from main but immediately get the response in form of a promise.

Long story short, my idea is the following: When the main process sends a message, it's basically fire and forget, whereas the renderer process will by default also fire and forget (as a state update will almost certainly generate a message that will update the renderer's state), but for certain use cases it's easier from a programming perspective to just `await ipc.get` instead of `ipc.send` and afterwards `ipc.once`. Furthermore, this layout will discourage using synchronous messages which are currently in operation in the app, and allow us to move away from this.

**Comment by pihentagy** [comment]
I know you have just rewritten FSAL, but do you think this one file is one note can scale? There are some other note taking apps, that tried that but failed.

**Comment by nathanlesage** [comment]
I understand your concerns, but basically what I mean by "refactoring" is simply to change the `dispatch` algorithm: Instead of dispatching the stuff internally in the IPC class, I want every class handle the events they are about directly. This will simply involve removing the `handler` and attaching event listeners in the classes themselves where the events are being handled!

The reason I opened up this issue is to have other people get involved because multiple pairs of eyes see more than one, and this way some potential bottlenecks could be spotted before we even begin work on this!
